llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp

```diff
@@ -1479,6 +1479,27 @@ void RISCVDAGToDAGISel::Select(SDNode *Node) {
           return;
         }
       }
+
+      // Look for (and (sra y, c2), c1) where c1 is a shifted mask with c3
+      // leading zeros and c4 trailing zeros. If c2 is greater than c3, we can
+      // use (slli (srli (srai y, c2 - c3), c3 + c4), c4).
+      if (isShiftedMask_64(C1) && !Skip) {
```

Comment:
The regression still exists. Can you add a test for `and (sra y, c2), -8`?

```suggestion
      if (isShiftedMask_64(C1) && !isInt<12>(C1)) {
```

Reply:
The one you put in #101571? Or a new one?

Reply:
The regression you put in #101751? Or a new one?

Reply:
It's not possible for the and immediate to be -8 here. It needs to have leading zeros.

Reply:
Reproducer:
```
define i64 @func0000000000000004(i64 %x) nounwind {
entry:
  %y = ashr i64 %x, 8
  %z = and i64 %y, -8
  ret i64 %z
}
```
Before:
```
func0000000000000004:
        srai    a0, a0, 8
        andi    a0, a0, -8
        ret
```
After this patch:
```
func0000000000000004:
        srai    a0, a0, 8
        srli    a0, a0, 3
        slli    a0, a0, 3
        ret
```

>  It needs to have leading zeros.

But it is not guarded by `isShiftedMask_64`. We should explicitly check this corner case.